crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -678,9 +699,31 @@ fn handle_own_line_comment_after_branch<'a>(
     preceding: AnyNodeRef<'a>,
     source: &str,
 ) -> CommentPlacement<'a> {
-    let Some(last_child) = preceding.last_child_in_body() else {
-        return CommentPlacement::Default(comment);
-    };
+    // If the preceding node has a body, we want the last child - e.g.
+    //
+    // ```python
+    // if True:
+    //     def foo():
+    //         something
+    //         last_child
+    //             # comment
+    // else:
+    //     pass
+    // ```
+    //
+    // Otherwise, the preceding node is the last statement in the body
+    // of the preceding branch, in which case we can take it as our
+    // `last_child` here - e.g.
+    //
+    // ```python
+    // if True:
+    //     something
+    //     last_child
+    //         # comment
+    // else:
+    //     pass
+    // ```
+    let last_child = preceding.last_child_in_body().unwrap_or(preceding);
```

Comment:
This seems risky. There isn't really anything enforcing that we're within a block statement. I think we, at least, want to assert that the enclosing node is a block statement.